### DISCUSSION TITLE
logging using RichHandler can't save teminal output in the hydra.log file

### DISCUSSION THREAD

**Discussion by QingYunA** [facebookresearch/hydra#2804] [Q&A] [answered]
Due to the good feature in the [rich package](https://rich.readthedocs.io/en/latest/logging.html), i used `RichHandler` to replace the default handler in `logging`, however the info ouputs by `logging(RichHandler)` can't be recorded in the hydra.log.

the code below is the example code which can easily reproduce
```python
import hydra
from omegaconf import DictConfig, OmegaConf
import logging
from rich.logging import RichHandler

log = logging.getLogger(__name__)
FORMAT = "%(message)s"
log.setLevel(logging.DEBUG)
log.addHandler(RichHandler())
log.propagate = True


def log_myinfo():
    log.info("This is a info message")
    log.debug("This is a debug message")
    log.warning("This is a warning message")
    log.error("This is a error message")
    log.critical("This is a critical message")


@hydra.main(version_base="1.3", config_path="./conf", config_name="test.yaml")
def my_app(cfg: DictConfig) -> None:
    log_myinfo()
    log.propagate = False
    log_myinfo()


if __name__ == "__main__":
    my_app()
```
the `test.yaml` is very simple which just contain one line `a: 1`.

first  i set `log.propagate=True` , in my terminal i can see Rich outputs and logging's parents logging:
![图片](https://github.com/facebookresearch/hydra/assets/41177312/2e8cd4b1-a043-4ee2-ab40-e0356301e0ee)
then i set `log.propagate=False`, which will not call the parent logging.
terminal outputs:
![图片](https://github.com/facebookresearch/hydra/assets/41177312/3d5eb119-e188-4ccf-84df-49f7fd2cfe1c)

but in the hydra_test.log, logging used Richhandler can't be saved.
![图片](https://github.com/facebookresearch/hydra/assets/41177312/c038e540-ab6d-4b91-b384-c1b89f9029ac)

What can i do to solve this problem?

**Comment by odelalleau** [comment]
I'm not 100% sure, but there's a good chance this may be related to #2765. You could try to adapt some of the suggestions in that issue to RichHandler.

**Comment by QingYunA** [comment, marked as answer]
Thx!
Inspired by #2765, by easily revise my code, I replace the original .log file by logging.FileHandler.
The code bellow is a pratically function to generate a logger.
```python
import logging
from rich.logging import RichHandler
import os

def get_logger():
    hydra_path = hydra.core.hydra_config.HydraConfig.get().runtime.output_dir
    job_name = hydra.core.hydra_config.HydraConfig.get().job.name

    file_handler = logging.FileHandler(os.path.join(hydra_path, f"{job_name}.log"))
    rich_handler = RichHandler()

    log = logging.getLogger(__name__)
    log.setLevel(logging.DEBUG)
    log.addHandler(rich_handler)
    log.addHandler(file_handler)
    log.propagate = False
    log.info("Successfully create rich logger")

    return log
```
Hope I can helps guys like me.

**Reply by odelalleau** [reply]
Thanks for sharing your solution!
